File: app/src/main/java/com/balicodes/quicksms/SMSFormFragment.java
package com.balicodes.quicksms;

import android.app.AlertDialog;
import android.content.ContentUris;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.database.Cursor;
import android.database.DataSetObserver;
import android.net.Uri;
import android.os.Bundle;
import android.provider.ContactsContract;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.text.TextUtils;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.Switch;
import android.widget.TextView;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by eka on 6/27/15.
 */
public class SMSFormFragment extends Fragment {

    private Button saveBtn;
    private EditText smstitle, message;
    private Switch addShortcut;
    private TextView addRecipient;
    private List<String[]> recipients = new ArrayList<String[]>();
    private NoScrollListView recipientListView;
    private RecipientListAdapter recipientListAdapter;
    private DBHelper dbHelper;
    private SMSItem smsItem;
    private int recipientPickIndex;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {

        // Contact data request
        if (requestCode == Config.PICK_CONTACT_REQUEST && resultCode == getActivity().RESULT_OK) {

            String phoneNo = null;
            String displayName = null;

            Uri uri = data.getData();
            Cursor cursor = getActivity().getContentResolver().query(uri, null, null, null, null);
            cursor.moveToFirst();

            int phoneIndex = cursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.NUMBER);
            int displayNameIndex = cursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME);
            phoneNo = cursor.getString(phoneIndex);
            displayName = cursor.getString(displayNameIndex);
            String[] recipient = new String[]{displayName, phoneNo};

            recipientListAdapter.addContactToItem(recipientPickIndex, recipient);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.sms_form_fragment, container, false);
        smstitle = (EditText) view.findViewById(R.id.titleTxt);
        message = (EditText) view.findViewById(R.id.messageTxt);

        recipientListView = (NoScrollListView) view.findViewById(R.id.receiverListView);
        recipientListAdapter = new RecipientListAdapter(this, recipients);
        recipientListView.setAdapter(recipientListAdapter);
        recipientListView.setDivider(null);
        recipientListView.setExpanded(true);

        addRecipient = (Button) view.findViewById(R.id.addReceiverBtn);
        addRecipient.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (recipients.size() < Config.MAX_RECIPIENTS_PER_SMS) {
                    saveRecipientsState(); // before adding new recipient, make sure we save any changes.
                    recipientListAdapter.addItem(new String[]{"", ""});
                } else {
                    String msg = String.format(getActivity().getResources().getString(R.string.max_recipients_warning), String.valueOf(Config.MAX_RECIPIENTS_PER_SMS));
                    Toast.makeText(getActivity(), msg, Toast.LENGTH_LONG).show();
                }
            }
        });

        addShortcut = (Switch) view.findViewById(R.id.addShortcut);

        saveBtn = (Button) view.findViewById(R.id.saveBtn);
        saveBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                saveSMS();
            }
        });

        // Show hide addRecipient
        recipientListAdapter.registerDataSetObserver(new DataSetObserver() {
            @Override
            public void onChanged() {
                if(recipients.size() < Config.MAX_RECIPIENTS_PER_SMS){
                    addRecipient.setVisibility(View.VISIBLE);
                }else{
                    addRecipient.setVisibility(View.GONE);
                }
            }
        });

        return view;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        ((MainActivity) getActivity()).getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        Bundle bundle = getArguments();
        // edit mode
        if (bundle != null) {
            smsItem = SMSItem.fromBundle(bundle);

            smstitle.setText(smsItem.getTitle());
            message.setText(smsItem.getMessage());
            recipients.addAll(SMSItem.parseReceiverCSV(smsItem.getNumber()));
            addShortcut.setChecked("YES".equals(smsItem.getShortcut()));

            // new mode
        } else {
            smsItem = new SMSItem();
            recipients.add(new String[]{"", ""});
        }
        recipientListAdapter.notifyDataSetChanged();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        if (smsItem.getId() > 0) {
            inflater.inflate(R.menu.sms_form_menu, menu);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();

        switch (id) {
            case R.id.action_delete:
                deleteSMS();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    private void hideKeyboard() {
        // Check if no view has focus:
        View view = getActivity().getCurrentFocus();
        if (view != null) {
            InputMethodManager inputManager = (InputMethodManager) getActivity().getSystemService(Context.INPUT_METHOD_SERVICE);
            inputManager.hideSoftInputFromWindow(view.getWindowToken(), InputMethodManager.HIDE_NOT_ALWAYS);
        }
    }

    private void deleteSMS() {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        builder.setTitle(R.string.confirm_delete_title);

        builder.setPositiveButton(R.string.yes, new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int id) {
                toggleShortcut(false); // delete shortcut
                smsItem.delete(SMSFormFragment.this.getActivity());

                Toast.makeText(getActivity(), R.string.message_deleted, Toast.LENGTH_SHORT).show();
                getActivity().onBackPressed();
            }
        });
        builder.setNegativeButton(R.string.cancel, new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int id) {
            }
        });
        AlertDialog dialog = builder.create();
        dialog.show();
    }

    /* This is to store any changes made to the recipients input back to the `recipients` */
    private void saveRecipientsState() {
        for (int i = 0; i < recipients.size(); i++) {
            View v = recipientListView.getChildAt(i);
            String name = ((EditText) v.findViewById(R.id.recName)).getText().toString().trim();
            String number = ((EditText) v.findViewById(R.id.recNumber)).getText().toString().trim();
            recipients.set(i, new String[]{name, number});
        }
    }

    private void saveSMS() {
        // do some simple validations
        if (smstitle.getText().toString().isEmpty()) {
            smstitle.setError(getString(R.string.title_error));
            return;
        }

        // get recipient as csv
        List<String> rec = new ArrayList<String>();
        for (int i = 0; i < recipients.size(); i++) {
            View v = recipientListView.getChildAt(i);
            String name = ((EditText) v.findViewById(R.id.recName)).getText().toString().trim();
            String number = ((EditText) v.findViewById(R.id.recNumber)).getText().toString().trim();
            if (number.length() > 0) {
                rec.add(name + ":" + number);
            }
        }
        String csv = TextUtils.join(",", rec);

        if (csv.isEmpty()) {
            Toast.makeText(getActivity(), R.string.number_error, Toast.LENGTH_LONG).show();
            return;
        }
        if (message.getText().toString().isEmpty()) {
            message.setError(getString(R.string.message_error));
            return;
        }

        String shortcut = addShortcut.isChecked() ? "YES" : "NO";

        if (smsItem.getId() > 0) {
            // Update
            smsItem.update(
                    getActivity(),
                    smstitle.getText().toString(),
                    csv,
                    message.getText().toString(),
                    shortcut);
        } else {
            // Create
            smsItem = SMSItem.create(
                    getActivity(),
                    smstitle.getText().toString(),
                    csv,
                    message.getText().toString(),
                    shortcut);
        }
        toggleShortcut(addShortcut.isChecked());

        hideKeyboard();
        Toast.makeText(getActivity(), R.string.message_saved, Toast.LENGTH_SHORT).show();
        getActivity().onBackPressed();
    }

    // used to set reference to recipient item index before contact Pick.
    public void setRecipientPickIndex(int index) {
        recipientPickIndex = index;
    }

    private void toggleShortcut(boolean create) {
        Intent shortCutInt = new Intent(getActivity().getApplicationContext(), ShortcutHandlerActivity.class);
        shortCutInt.setAction(Intent.ACTION_MAIN);
        shortCutInt.setData(ContentUris.withAppendedId(Uri.parse(Config.SMS_DATA_BASE_URI), smsItem.getId()));

        Intent addInt = new Intent();
        addInt.putExtra(Intent.EXTRA_SHORTCUT_INTENT, shortCutInt);
        addInt.putExtra(Intent.EXTRA_SHORTCUT_NAME, smsItem.getTitle());
        addInt.putExtra("duplicate", false);
        addInt.putExtra(Intent.EXTRA_SHORTCUT_ICON_RESOURCE,
                Intent.ShortcutIconResource.fromContext(getActivity().getApplicationContext(),
                        R.drawable.ic_launcher));
        if (create) {
            addInt.setAction("com.android.launcher.action.INSTALL_SHORTCUT");
        } else {
            addInt.setAction("com.android.launcher.action.UNINSTALL_SHORTCUT");
        }

        getActivity().getApplicationContext().sendBroadcast(addInt);
    }
}

File: app/src/main/java/com/balicodes/quicksms/SMSFormFragment.kt
/*
 * Copyright (C) 2016-2018 Eka Putra
 *
 * Quick SMS is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.balicodes.quicksms

import android.app.Activity.RESULT_OK
import android.app.AlertDialog
import android.app.PendingIntent
import android.arch.lifecycle.ViewModelProviders
import android.content.ContentUris
import android.content.Context
import android.content.Intent
import android.content.pm.ShortcutInfo
import android.content.pm.ShortcutManager
import android.database.DataSetObserver
import android.graphics.drawable.Icon
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.provider.ContactsContract
import android.support.v4.app.Fragment
import android.text.TextUtils
import android.view.*
import android.view.inputmethod.InputMethodManager
import android.widget.*
import com.balicodes.quicksms.entity.MessageEntity
import com.balicodes.quicksms.model.SMSItem
import com.balicodes.quicksms.viewmodel.MessageViewModel

class SMSFormFragment : Fragment() {

    lateinit var viewModel: MessageViewModel
    lateinit var recipientListView: NoScrollListView
    lateinit var smstitle: EditText
    lateinit var message: EditText
    lateinit var addShortcut: Switch
    lateinit var addRecipient: TextView

    private val recipients = ArrayList<Array<String>>()

    private var recipientListAdapter: RecipientListAdapter? = null
    private var smsItem: SMSItem? = null
    private var recipientPickIndex: Int = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)

        (activity as MainActivity).supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        viewModel = ViewModelProviders.of(requireActivity()).get(MessageViewModel::class.java)
        viewModel.getSelectedMessage().observe(this, android.arch.lifecycle.Observer { setFormData(it) })
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {

        // Contact data request
        if (requestCode == Config.PICK_CONTACT_REQUEST && resultCode == RESULT_OK && data != null) {

            val phoneNo: String
            val displayName: String

            val uri = data.data
            val cursor = requireActivity().contentResolver.query(uri, null, null, null, null)
            cursor.moveToFirst()

            val phoneIndex = cursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.NUMBER)
            val displayNameIndex = cursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME)
            phoneNo = cursor.getString(phoneIndex)
            displayName = cursor.getString(displayNameIndex)
            val recipient = arrayOf(displayName, phoneNo)

            cursor.close()

            recipientListAdapter!!.addContactToItem(recipientPickIndex, recipient)
        }
    }


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.sms_form_fragment, container, false)
        smstitle = view.findViewById(R.id.titleTxt)
        message = view.findViewById(R.id.messageTxt)

        recipientListView = view.findViewById(R.id.receiverListView)
        recipientListAdapter = RecipientListAdapter(this, recipients)
        recipientListView.adapter = recipientListAdapter
        recipientListView.divider = null
        recipientListView.expanded = true

        addRecipient = view.findViewById<Button>(R.id.addReceiverBtn)
        addRecipient.setOnClickListener {
            if (recipients.size < Config.MAX_RECIPIENTS_PER_SMS) {
                saveRecipientsState() // before adding new recipient, make sure we save any changes.
                recipientListAdapter!!.addItem(arrayOf("", ""))
            } else {
                val msg = String.format(requireActivity().resources.getString(R.string.max_recipients_warning), Config.MAX_RECIPIENTS_PER_SMS.toString())
                Toast.makeText(activity, msg, Toast.LENGTH_LONG).show()
            }
        }

        addShortcut = view.findViewById(R.id.addShortcut)

        val saveBtn = view.findViewById<Button>(R.id.saveBtn)
        saveBtn.setOnClickListener { saveMessage() }

        // Show hide addRecipient
        recipientListAdapter!!.registerDataSetObserver(object : DataSetObserver() {
            override fun onChanged() {
                if (recipients.size < Config.MAX_RECIPIENTS_PER_SMS) {
                    addRecipient.visibility = View.VISIBLE
                } else {
                    addRecipient.visibility = View.GONE
                }
            }
        })

        return view
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        super.onCreateOptionsMenu(menu, inflater)

        viewModel.getSelectedMessage().observe(this, android.arch.lifecycle.Observer {
            it?.let { inflater?.inflate(R.menu.sms_form_menu, menu) }
        })
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        val id = item!!.itemId

        when (id) {
            R.id.action_delete -> deleteMessage()
            R.id.action_copy -> viewModel.copyMessage(smsItem!!.toEntity(), this::afterDuplicate)
        }
        return super.onOptionsItemSelected(item)
    }

    private fun setFormData(messageEntity: MessageEntity?) {
        messageEntity?.let {
            smsItem = it.toSmsItem()

            smstitle.setText(it.title)
            message.setText(it.message)
            recipients.addAll(SMSItem.parseReceiverCSV(it.number))
            addShortcut.isChecked = (it.addShortcut == SMSItem.SHORTCUT_YES)

            recipientListAdapter!!.notifyDataSetChanged()
        }
    }

    private fun hideKeyboard() {
        // Check if no view has focus:
        val view = requireActivity().currentFocus
        if (view != null) {
            val inputManager = requireContext().getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            inputManager.hideSoftInputFromWindow(view.windowToken, InputMethodManager.HIDE_NOT_ALWAYS)
        }
    }

    private fun deleteMessage() {
        val builder = AlertDialog.Builder(requireActivity())
        builder.setTitle(R.string.confirm_delete_title)

        builder.setPositiveButton(R.string.yes) { _, _ ->
            viewModel.deleteMessage(smsItem!!.toEntity(), this::afterDelete)
        }
        builder.setNegativeButton(R.string.cancel) { _, _ -> }
        val dialog = builder.create()
        dialog.show()
    }

    /* This is to store any changes made to the recipients input back to the `recipients` */
    private fun saveRecipientsState() {
        for (i in recipients.indices) {
            val v = recipientListView.getChildAt(i)
            val name = (v.findViewById<View>(R.id.recName) as EditText).text.toString().trim { it <= ' ' }
            val number = (v.findViewById<View>(R.id.recNumber) as EditText).text.toString().trim { it <= ' ' }
            recipients[i] = arrayOf(name, number)
        }
    }

    // Save / Update the message.
    private fun saveMessage() {
        // do some simple validations
        if (smstitle.text.toString().isEmpty()) {
            smstitle.error = getString(R.string.title_error)
            return
        }

        if (message.text.toString().isEmpty()) {
            message.error = getString(R.string.message_error)
            return
        }
        // get recipient as csv
        val rec = ArrayList<String>()
        for (i in recipients.indices) {
            val v = recipientListView.getChildAt(i)
            val name = (v.findViewById<View>(R.id.recName) as EditText).text.toString().trim { it <= ' ' }
            val number = (v.findViewById<View>(R.id.recNumber) as EditText).text.toString().trim { it <= ' ' }
            if (number.isNotBlank()) {
                rec.add(name + ":" + number)
            }
        }
        val csv = TextUtils.join(",", rec)

        if (csv.isEmpty()) {
            Toast.makeText(requireActivity(), R.string.number_error, Toast.LENGTH_LONG).show()
            return
        }

        val shortcut = if (addShortcut.isChecked) SMSItem.SHORTCUT_YES else SMSItem.SHORTCUT_NO

        if (smsItem != null) {
            // Update
            val item = SMSItem(smsItem!!.id, smstitle.text.toString(), csv, message.text.toString(), shortcut)
            viewModel.updateMessage(item.toEntity(), this::afterSave)

        } else {
            // Create
            val item = SMSItem(0L, smstitle.text.toString(), csv, message.text.toString(), shortcut)
            viewModel.insertMessage(item.toEntity(), this::afterSave)
        }
    }

    // After message has been saved.
    private fun afterSave(messageEntity: MessageEntity) {
        toggleShortcut(messageEntity, messageEntity.addShortcut == SMSItem.SHORTCUT_YES)
        hideKeyboard()

        Toast.makeText(requireActivity(), R.string.message_saved, Toast.LENGTH_SHORT).show()
        requireActivity().onBackPressed()
    }

    // After message has been duplicated.
    private fun afterDuplicate(messageEntity: MessageEntity) {
        // Clear the recipients and notify adapter
        recipients.clear()
        recipientListAdapter?.notifyDataSetChanged()

        // Set selected message.
        viewModel.selectMessage(messageEntity)

        // POP current backstack
        requireActivity().supportFragmentManager.popBackStack()
        val duplicateForm = SMSFormFragment()

        // Replace current fragment with the new one.
        val ft = requireActivity().supportFragmentManager.beginTransaction()
        ft.replace(R.id.container, duplicateForm)
        ft.addToBackStack("sms_form")
        ft.commit()

        Toast.makeText(requireActivity(), R.string.message_copied, Toast.LENGTH_SHORT).show()
    }

    // After message deleted.
    private fun afterDelete(messageEntity: MessageEntity) {
        toggleShortcut(messageEntity, false) // delete shortcut

        Toast.makeText(requireActivity(), R.string.message_deleted, Toast.LENGTH_SHORT).show()
        requireActivity().onBackPressed()
    }

    // used to set reference to recipient item index before contact Pick.
    fun setRecipientPickIndex(index: Int) {
        recipientPickIndex = index
    }

    // Toogle shortcut icon
    // TODO: Move this code block into its own class.
    private fun toggleShortcut(messageEntity: MessageEntity, create: Boolean) {

        val shortCutInt = Intent(requireContext(), ShortcutHandlerActivity::class.java)
        shortCutInt.action = Intent.ACTION_MAIN
        shortCutInt.data = ContentUris.withAppendedId(Uri.parse(Config.SMS_DATA_BASE_URI), messageEntity.id!!)

        // On Android 25+ we can create dynamic shortcuts
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N_MR1) {
            val shortcutManager = requireContext().getSystemService(ShortcutManager::class.java)
            val shortcutId = "QSMS_".format(messageEntity.id)

            val shortcutInfo = ShortcutInfo.Builder(requireActivity(), shortcutId)
                    .setShortLabel(messageEntity.title)
                    .setIcon(Icon.createWithResource(requireActivity(), R.drawable.ic_launcher_shortcut))
                    .setIntent(shortCutInt)
                    .build()

            // On Android 26+ we can create pinned shortcut on home screen.
            // else, we only create dynamic shortcut.
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && shortcutManager.isRequestPinShortcutSupported) {
                val shortcuts = shortcutManager.pinnedShortcuts

                var exists = false
                loop@ for (s in shortcuts) {
                    if (s.id == shortcutId) {
                        exists = true
                        break@loop
                    }
                }
                if (create && !exists) {
                    val pinnedShortcutCallbackIntent = shortcutManager.createShortcutResultIntent(shortcutInfo)
                    val successCallback = PendingIntent.getBroadcast(requireContext(), 0, pinnedShortcutCallbackIntent, 0)
                    shortcutManager.requestPinShortcut(shortcutInfo, successCallback.intentSender)
                } else if (create && exists) {
                    shortcutManager.updateShortcuts(listOf(shortcutInfo))
                } else {
                    val builder = AlertDialog.Builder(requireActivity())
                    builder.setTitle("Deleting shortcut")
                    builder.setMessage("Unfortunately Android doesn't allow us to delete the shortcut for you, so you have to delete it manually :)")
                    builder.setPositiveButton(R.string.yes) { _, _ -> }
                    val dialog = builder.create()
                    dialog.show()
                }

            } else {
                val shortcuts = shortcutManager.dynamicShortcuts

                var exists = false
                loop@ for (s in shortcuts) {
                    if (s.id == shortcutId) {
                        exists = true
                        break@loop
                    }
                }
                if (create && !exists) {
                    shortcutManager.addDynamicShortcuts(listOf(shortcutInfo))
                } else if (create && exists) {
                    shortcutManager.updateShortcuts(listOf(shortcutInfo))
                } else {
                    shortcutManager.removeDynamicShortcuts(listOf(shortcutId))
                }
            }

        } else {
            val addInt = Intent()
            addInt.putExtra(Intent.EXTRA_SHORTCUT_INTENT, shortCutInt)
            addInt.putExtra(Intent.EXTRA_SHORTCUT_NAME, messageEntity.title)
            addInt.putExtra("duplicate", false)
            addInt.putExtra(Intent.EXTRA_SHORTCUT_ICON_RESOURCE,
                    Intent.ShortcutIconResource.fromContext(requireActivity().applicationContext,
                            R.drawable.ic_launcher_shortcut))
            if (create) {
                addInt.action = "com.android.launcher.action.INSTALL_SHORTCUT"
            } else {
                addInt.action = "com.android.launcher.action.UNINSTALL_SHORTCUT"
            }

            requireContext().applicationContext.sendBroadcast(addInt)
        }
    }
}

